### PR TITLE
Feature post api for purchase

### PR DESCRIPTION
**Pull request by vivekhit2000**

### CODE REVIEW

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        saveRepository(purchaseObject);
+        return PurchaseResponse.purchaseBuilder().success(true).id(purchaseObject.getId()).userName(purchaseObject.getUserName()).product(purchaseObject.getProduct()).amount(purchaseObject.getAmount()).date(purchaseObject.getDate()).build();
+    }
+
+    /**
+     * Calling purchase repository to store the valid user information to the database
+     * @param purchase
+     * @return
+     */
+    public Purchase saveRepository(Purchase purchase) {
```
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: why is this method public

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    public PurchaseService(PurchaseRepository purchaseRepository) {
+        this.purchaseRepository = purchaseRepository;
+    }
+
+    private static Purchase createPurchaseObjectFromRequest(PurchaseRequest purchaseRequest) {
+        return Purchase.builder().userName(purchaseRequest.getUserName()).product(purchaseRequest.getProduct()).amount(purchaseRequest.getAmount()).date(purchaseRequest.getDate()).build();
+    }
+
+    public PurchaseResponse createValidPurchase(PurchaseRequest purchaseRequestObject) {
```
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: Are you creating another method with name createInvalidPurchase()? 
If not just createPurchase() is a good enough name

Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: Updated

File: src/main/java/com/elixr/poc/service/PurchaseService.java (hunk 1 above)
Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: I make it private now

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+@Service
+public class PurchaseService {
+    private final PurchaseRepository purchaseRepository;
+
+    public PurchaseService(PurchaseRepository purchaseRepository) {
+        this.purchaseRepository = purchaseRepository;
+    }
+
+    private static Purchase createPurchaseObjectFromRequest(PurchaseRequest purchaseRequest) {
```
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: Why is this method static?

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        saveRepository(purchaseObject);
+        return PurchaseResponse.purchaseBuilder().success(true).id(purchaseObject.getId()).userName(purchaseObject.getUserName()).product(purchaseObject.getProduct()).amount(purchaseObject.getAmount()).date(purchaseObject.getDate()).build();
+    }
+
+    /**
+     * Calling purchase repository to store the valid user information to the database
+     * @param purchase
+     * @return
+     */
+    private Purchase saveRepository(Purchase purchase) {
```
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: saveRepository name suggests you are saving respository object which is wrong.
either rename it as savePurchaseInDB or savePurchase

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    private static Purchase createPurchaseObjectFromRequest(PurchaseRequest purchaseRequest) {
+        return Purchase.builder().userName(purchaseRequest.getUserName()).product(purchaseRequest.getProduct()).amount(purchaseRequest.getAmount()).date(purchaseRequest.getDate()).build();
+    }
+
+    public PurchaseResponse createPurchase(PurchaseRequest purchaseRequestObject) {
+        Purchase purchaseObject = createPurchaseObjectFromRequest(purchaseRequestObject);
+        saveRepository(purchaseObject);
+        return PurchaseResponse.purchaseBuilder().success(true).id(purchaseObject.getId()).userName(purchaseObject.getUserName()).product(purchaseObject.getProduct()).amount(purchaseObject.getAmount()).date(purchaseObject.getDate()).build();
```
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: Format this line to make readable

File: src/main/java/com/elixr/poc/service/PurchaseService.java (hunk 3 above)
Comment (mh-sraveendran) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: Service should not know anything about rest objects. 
createPurchaseObjectFromRequest() should be done inside controller and purchase object should be passed

Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: Updated

Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: removed  static

File: src/main/java/com/elixr/poc/service/PurchaseService.java (hunk 4 above)
Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: Updated

File: src/main/java/com/elixr/poc/service/PurchaseService.java (hunk 5 above)
Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: updated

File: src/main/java/com/elixr/poc/rest/controller/PurchaseCreationController.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+public class PurchaseCreationController {
+
+    private final PurchaseService purchaseService;
+
+    public PurchaseCreationController(PurchaseService purchaseService) {
+        this.purchaseService = purchaseService;
+    }
+
+    @PostMapping("/purchase")
+    public ResponseEntity<?> addPurchase(@RequestBody @Valid PurchaseRequest newPurchase) {
```
Comment (mh-spunnappala) [inline src/main/java/com/elixr/poc/rest/controller/PurchaseCreationController.java]: Remove usage of wildcard type

File: src/main/java/com/elixr/poc/rest/controller/PurchaseCreationController.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        Purchase purchase = Purchase.builder().userName(newPurchase.getUserName()).product(newPurchase.getProduct()).amount(newPurchase.getAmount()).date(newPurchase.getDate()).build();
+        purchaseService.createPurchase(purchase);
+
+        return new ResponseEntity<>(PurchaseResponse.purchaseBuilder().success(true).id(purchase.getId()).userName(purchase.getUserName()).product(purchase.getProduct()).amount(purchase.getAmount()).date(purchase.getDate()).build(), HttpStatus.OK);
```
Comment (mh-spunnappala) [inline src/main/java/com/elixr/poc/rest/controller/PurchaseCreationController.java]: format the method

File: src/main/java/com/elixr/poc/rest/response/PurchaseResponse.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+public class PurchaseResponse extends ErrorResponse {
+
+    @Id
+    private UUID id;
+    private String userName;
+    private String product;
+    private String amount;
+    private String date;
+
+    @Builder(builderMethodName = "purchaseBuilder")
```
Comment (mh-spunnappala) [inline src/main/java/com/elixr/poc/rest/response/PurchaseResponse.java]: Use Lombok annotation for all argument constructor

File: src/main/java/com/elixr/poc/service/PurchaseService.java [hunk 9]
```diff
@@ -15,6 +20,25 @@ public PurchaseService(PurchaseRepository purchaseRepository) {
...
+     * @param purchase
+     * @return
+     */
+    private Purchase savePurchase(Purchase purchase) {
+        if (purchase.getId() == null || purchase.getId().toString().isEmpty()) {
+            purchase.setId(UUID.randomUUID());
+        }
+        purchase = this.purchaseRepository.save(purchase);
+        return purchase;
+        this.purchaseRepository = purchaseRepository;
```
Comment (mh-spunnappala) [inline src/main/java/com/elixr/poc/service/PurchaseService.java]: why this assignment

**Review by mh-spunnappala — CHANGES_REQUESTED** [review]

File: src/main/java/com/elixr/poc/rest/response/PurchaseResponse.java (hunk 8 above)
Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/rest/response/PurchaseResponse.java]: updated

File: src/main/java/com/elixr/poc/service/PurchaseService.java (hunk 9 above)
Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/service/PurchaseService.java]: now, removed.

File: src/main/java/com/elixr/poc/rest/controller/PurchaseCreationController.java (hunk 7 above)
Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/rest/controller/PurchaseCreationController.java]: formatted.

File: src/main/java/com/elixr/poc/rest/controller/PurchaseCreationController.java (hunk 6 above)
Comment (vivekhit2000) [reply src/main/java/com/elixr/poc/rest/controller/PurchaseCreationController.java]: removed

**Pull request merged**